### PR TITLE
feat(init-reset): implemented small state machine for link opening and reset

### PR DESCRIPTION
**Pull request by gabrik**
Initializaiton message exchange from cold start

1) Connect side sends a message with empty payload and I flag set.
2) Listen side sends a message with empty payload and I+A flags set.
3) Data can be exchanged from this point on.

```
///  ┌─────────┐                ┌──────────┐   
///  │ Listen  │                │ Connect  │   
///  └────┬────┘                └────┬─────┘   
///       │                          │         
///       │       Init               │         
///       │◄─────────────────────────┤         
///       │                          │         
///       │        Init + Ack        │         
///       ├─────────────────────────►│         
///       │       Data               │         
///       │◄─────────────────────────┤         
///       ├─────────────────────────►│         
///       │                          │    
```
 If connect sides restarts:
1) Connect side sends a message with empty payload and I flag set.   
2) Listen side sends a message with empty payload and R flag set.
3) Connect side waits and goes back to cold start case.

```
///   ┌─────────┐                ┌──────────┐
///   │ Listen  │                │ Connect  │
///   └────┬────┘                └────┬─────┘
///        │        Init              │       
///        │◄─────────────────────────┤       
///        │       Reset              │       
///        ├─────────────────────────►│       
///        │                          │       
///        │                          │       
///        │                          │       
///        │        Init              │       
///        │◄─────────────────────────┤       
///        │       Init + Ack         │       
///        ├─────────────────────────►│       
///        │         Data             │       
///        │◄─────────────────────────┤       
///        ├─────────────────────────►│       
///        │                          │       
///        │                          │  
```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -235,9 +345,82 @@ impl ZSerial {
...
+            let (_read, hdr) = self.internal_read(&mut buff).await?;
+            log::trace!("Received header: {hdr:02X?}");
+
+            if hdr.has_a_flag() && hdr.has_i_flag() {
+                // correct initialiation
+                self.status = Status::Initialized;
+                break;
+            } else if hdr.has_r_flag() {
+                // we received a reset we must resend the init message after a small sleep
+                tokio::time::sleep(Duration::from_micros(SERIAL_CONNECT_THROTTLE_TIME_US)).await;
```
Comment (Mallets) [inline src/lib.rs]: Is this sleep really necessary?

**Comment by ur8us** [comment]
Guys, I am sorry for the headache, but it is clearly seen from the diagram that the behavior of the Connect side does not depend on the state of the flags in the new header field (the "H" byte). If the connect request is successful, data transfer begins. If not, the Connect side keeps trying. In other words, the H byte is useless. :-(

File: src/lib.rs [hunk 2]
```diff
@@ -235,9 +345,78 @@ impl ZSerial {
...
+        self.clear()?;
+
+        loop {
+            // wait for an empty message with I flag
+
+            let (_read, hdr) = self.internal_read(&mut buff).await?;
+            log::trace!("Received header: {hdr:02X?}");
+            if hdr.has_i_flag() {
+                // we send back a message with both I and A flags
+                self.internal_write(&[0u8], Header::new(I_FLAG | A_FLAG))
```
Comment (sashacmc) [inline src/lib.rs]: It is not an empty payload, but payload with length 1, but according to the format we expected length 0

Comment (gabrik) [reply src/lib.rs]: Indeed, but we do not even check the payload. Its indeed a payload with lenght 1 and a single 0 byte.

Comment (sashacmc) [reply src/lib.rs]: We don't check the payload here, but in pico the message is read in its entirety, and for this a buffer is needed, for connection messages I would prefer not to allocate it.

Comment (gabrik) [reply src/lib.rs]: I'm going to fix it.

**Review by sashacmc — APPROVED** [review]

**Pull request merged**